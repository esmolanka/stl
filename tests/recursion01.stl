
type Peano = < Zero, Succ : Peano >
#eval Peano

type ConsList a = <Nil : Unit, Cons : {head+ : a, tail+ : ConsList}>
#eval ConsList
#eval ConsList Unit
#eval ConsList Peano

type ConsList2 a = <Nil : Unit, Cons : {head+ : a, tail+ : ConsList a}>
#eval ConsList2
#eval ConsList2 Unit
#eval ConsList2 Peano

#check ConsList Unit <: ConsList2 Unit
#check ConsList2 Unit <: ConsList Unit
