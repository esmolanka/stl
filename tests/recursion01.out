tests/recursion01.types:3:1-3:12:
  Normalised:
    μ Peano. ∀ (ρ : Row). <Zero? : Unit | Succ? : Peano | … ρ>
  Kind:
    Type
tests/recursion01.types:6:1-6:15:
  Normalised:
    λ a. μ ConsList. ∀ (ρ : Row).
      < Nil? : Unit
      | Cons? : ∃ (ρ : Row). {head : a, tail : ConsList, … ρ}
      | … ρ
      >
  Kind:
    Type → Type
tests/recursion01.types:7:1-7:20:
  Normalised:
    μ ConsList. ∀ (ρ : Row).
      < Nil? : Unit
      | Cons? : ∃ (ρ : Row). {head : Unit, tail : ConsList, … ρ}
      | … ρ
      >
  Kind:
    Type
tests/recursion01.types:8:1-8:21:
  Normalised:
    μ ConsList. ∀ (ρ : Row).
      < Nil? : Unit
      | Cons? :
          ∃ (ρ : Row).
            { head : μ Peano. ∀ (ρ : Row). <Zero? : Unit | Succ? : Peano | … ρ>
            , tail : ConsList
            , … ρ
            }
      | … ρ
      >
  Kind:
    Type
tests/recursion01.types:11:1-11:16:
  Normalised:
    λ a. ∀ (ρ : Row).
      < Nil? : Unit
      | Cons? :
          ∃ (ρ : Row).
            { head : a
            , tail :
                μ ConsList. ∀ (ρ : Row).
                  < Nil? : Unit
                  | Cons? : ∃ (ρ : Row). {head : a, tail : ConsList, … ρ}
                  | … ρ
                  >
            , … ρ
            }
      | … ρ
      >
  Kind:
    Type → Type
tests/recursion01.types:12:1-12:21:
  Normalised:
    ∀ (ρ : Row).
      < Nil? : Unit
      | Cons? :
          ∃ (ρ : Row).
            { head : Unit
            , tail :
                μ ConsList. ∀ (ρ : Row).
                  < Nil? : Unit
                  | Cons? : ∃ (ρ : Row). {head : Unit, tail : ConsList, … ρ}
                  | … ρ
                  >
            , … ρ
            }
      | … ρ
      >
  Kind:
    Type
tests/recursion01.types:13:1-13:22:
  Normalised:
    ∀ (ρ : Row).
      < Nil? : Unit
      | Cons? :
          ∃ (ρ : Row).
            { head : μ Peano. ∀ (ρ : Row). <Zero? : Unit | Succ? : Peano | … ρ>
            , tail :
                μ ConsList. ∀ (ρ : Row).
                  < Nil? : Unit
                  | Cons? :
                      ∃ (ρ : Row).
                        { head :
                            μ Peano. ∀ (ρ : Row).
                              <Zero? : Unit | Succ? : Peano | … ρ>
                        , tail : ConsList
                        , … ρ
                        }
                  | … ρ
                  >
            , … ρ
            }
      | … ρ
      >
  Kind:
    Type
tests/recursion01.types:15:1-15:39:
  Subtype:
    μ ConsList. ∀ (ρ : Row).
      < Nil? : Unit
      | Cons? : ∃ (ρ : Row). {head : Unit, tail : ConsList, … ρ}
      | … ρ
      >
  Supertype:
    ∀ (ρ : Row).
      < Nil? : Unit
      | Cons? :
          ∃ (ρ : Row).
            { head : Unit
            , tail :
                μ ConsList. ∀ (ρ : Row).
                  < Nil? : Unit
                  | Cons? : ∃ (ρ : Row). {head : Unit, tail : ConsList, … ρ}
                  | … ρ
                  >
            , … ρ
            }
      | … ρ
      >
  Result:
    OK
tests/recursion01.types:16:1-16:39:
  Subtype:
    ∀ (ρ : Row).
      < Nil? : Unit
      | Cons? :
          ∃ (ρ : Row).
            { head : Unit
            , tail :
                μ ConsList. ∀ (ρ : Row).
                  < Nil? : Unit
                  | Cons? : ∃ (ρ : Row). {head : Unit, tail : ConsList, … ρ}
                  | … ρ
                  >
            , … ρ
            }
      | … ρ
      >
  Supertype:
    μ ConsList. ∀ (ρ : Row).
      < Nil? : Unit
      | Cons? : ∃ (ρ : Row). {head : Unit, tail : ConsList, … ρ}
      | … ρ
      >
  Result:
    OK